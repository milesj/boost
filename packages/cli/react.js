module.exports = require('./cjs/react.cjs');
