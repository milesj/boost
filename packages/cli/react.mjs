export * from './mjs/react.mjs';
