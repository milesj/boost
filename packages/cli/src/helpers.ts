import { stripAnsi, style } from '@boost/terminal';
import { PrimitiveType } from '@boost/args/src';

export function applyMarkdown(text: string): string {
  return text.replace(/(?:\*|~|_){1,2}([^*~_]+)(?:\*|~|_){1,2}/giu, (match, body) => {
    if (match.startsWith('~')) {
      return style.strikethrough(body);
    } else if (match.startsWith('**') || match.startsWith('__')) {
      return style.bold(body);
    } else if (match.startsWith('*') || match.startsWith('_')) {
      return style.italic(body);
    }

    return match;
  });
}

export function formatValue(value: PrimitiveType): string {
  switch (typeof value) {
    case 'string':
      return `"${value}"`;
    default:
      return String(value);
  }
}

export function formatDescription(
  config: {
    choices?: PrimitiveType[];
    default?: PrimitiveType;
    description: string;
    deprecated?: boolean;
  },
  tags: string[] = [],
): string {
  let output = applyMarkdown(config.description);

  // Append choices after the description
  if (config.choices) {
    const choices = config.choices.map(
      choice => formatValue(choice) + (config.default === choice ? ' (default)' : ''),
    );

    output += ` (choices: ${choices.join(', ')})`;

    // Append default after description if no choices
  } else if (config.default) {
    output += ` (default: ${formatValue(config.default)})`;
  }

  // Tags go at the end of the description
  if (config.deprecated) {
    tags.unshift('deprecated');
  }

  if (tags.length > 0) {
    output += ' ';
    output += style.gray(`[${tags.join(', ')}]`);
  }

  return output;
}

export function formatType(
  config: {
    label?: string;
    multiple?: boolean;
    required?: boolean;
    type: string;
  },
  inline: boolean = false,
): string {
  let type = config.type + (config.multiple ? '[]' : '');

  if (inline) {
    if (config.label) {
      type = config.label;
    }

    type = config.required ? `<${type}>` : `[${type}]`;
  }

  return style.gray(type);
}

export function getLongestWidth(values: string[]): number {
  return values.reduce((sum, value) => {
    const text = stripAnsi(value);

    return text.length > sum ? text.length : sum;
  }, 0);
}
