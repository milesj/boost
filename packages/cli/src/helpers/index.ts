export * from './applyMarkdown';
export * from './applyStyle';
export * from './formatCommandCall';
export * from './formatDescription';
export * from './formatType';
export * from './formatValue';
export * from './getLongestWidth';
export * from './groupByCategory';
export * from './isArgvSize';
export * from './loadTheme';
export * from './mapCommandMetadata';
