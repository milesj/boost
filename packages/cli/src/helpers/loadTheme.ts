import { requireModule } from '@boost/common';
import { env } from '@boost/internal';
import { style } from '@boost/terminal';
import { CLIError } from '../CLIError';
import { ThemePalette } from '../types';

const loadedThemes = new Map<string, ThemePalette>();

export function loadTheme(): ThemePalette {
	const theme = env('CLI_THEME') ?? '';
	let palette = loadedThemes.get(theme);

	if (style.level > 0 && !palette && !!theme) {
		try {
			palette = requireModule(`@boost/theme-${theme}`);
		} catch {
			try {
				palette = requireModule(`boost-theme-${theme}`);
			} catch {
				throw new CLIError('THEME_UNKNOWN', [theme]);
			}
		}

		if (palette) {
			loadedThemes.set(theme, palette);
		}
	}

	// ANSI escapes + hexcodes dont work too well with snapshots
	const isTest = process.env.NODE_ENV === 'test';

	return {
		default: 'white',
		failure: 'red',
		info: 'cyan',
		// Use a hexcode since it sometimes renders as gray
		inverted: isTest ? 'black' : '#000',
		muted: 'gray',
		notice: 'magenta',
		success: 'green',
		warning: 'yellow',
		...palette,
	};
}
