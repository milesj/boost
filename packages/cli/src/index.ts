/**
 * @copyright   2020, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

export * from './CLIError';
export * from './Command';
export * from './constants';
export { Config } from './decorators/Config';
export * from './helpers';
export * as Arg from './metadata/args';
export * from './middleware';
export * from './Program';
export * from './types';
