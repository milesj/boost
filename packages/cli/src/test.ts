/* eslint-disable no-param-reassign */

import React from 'react';
import { render } from 'ink';
import { env } from '@boost/internal';
import { mockLogger } from '@boost/log/test';
import { stripAnsi } from '@boost/terminal';
import { Command, INTERNAL_OPTIONS, INTERNAL_PARAMS, Program } from '.';

import type {
	ExitCode,
	GlobalOptions,
	PrimitiveType,
	ProgramOptions,
	ProgramStreams,
	TaskContext,
} from '.';

export class MockReadStream {
	isTTY = false;
}

export class MockWriteStream {
	append: boolean = false;

	columns: number = 80;

	output: string;

	constructor(append: boolean = false) {
		this.append = append;
		this.output = '';
	}

	write(string: string) {
		if (this.append) {
			this.output += string;
		} else {
			this.output = string;
		}
	}

	get(): string {
		return this.output;
	}

	on() {}

	off() {}
}

/**
 * Returns mocked `stderr`, `stdout`, and `stdin` streams that can be passed to a `Program`.
 * This does not mock all stream functionality, only those required by Boost and Ink.
 *
 * ```ts
 * import { mockStreams } from '@boost/cli/test';
 *
 * const streams = mockStreams();
 * ```
 */
export function mockStreams(append?: boolean): ProgramStreams {
	return {
		stderr: new MockWriteStream(append),
		stdin: new MockReadStream(),
		stdout: new MockWriteStream(append),
	} as unknown as ProgramStreams;
}

/**
 * Returns a `Program` instance with required options pre-filled and streams mocked
 * (unless manually provided).
 *
 * ```ts
 * import { mockProgram } from '@boost/cli/test';
 *
 * const program = mockProgram({ name: 'Example' });
 * ```
 */
export function mockProgram(options?: Partial<ProgramOptions>, streams?: ProgramStreams): Program {
	return new Program(
		{
			bin: 'test',
			name: 'Test',
			version: '0.0.0',
			...options,
		},
		streams ?? mockStreams(),
	);
}

/**
 * Can be used to render a React component with Ink and return the rendered result
 * as a terminal compatible string. If `stripped` is true, it will strip ANSI
 * escape escape sequences.
 *
 * ```tsx
 * import { renderComponent } from '@boost/cli/test';
 * import TestComponent from '../src/components/TestComponent';
 *
 * it('renders a component', async () => {
 * 	expect(await renderComponent(<TestComponent />)).toMatchSnapshot();
 * });
 * ```
 *
 * > As an alternative, we also suggest using the official
 * > [ink-testing-library](https://github.com/vadimdemedes/ink-testing-library).
 */
export async function renderComponent(
	element: React.ReactElement,
	stripped: boolean = false,
): Promise<string> {
	let output = '';

	try {
		const { render: renderAsTest } = await import('ink-testing-library');
		const { lastFrame } = await renderAsTest(element);

		output = lastFrame() ?? '';
	} catch {
		const stdout = new MockWriteStream();

		await render(element, {
			debug: true,
			experimental: true,
			stdout: stdout as unknown as NodeJS.WriteStream,
		});

		output = stdout.get();
	}

	return stripped ? stripAnsi(output) : output;
}

/**
 * Runs a `Command` outside the context of a `Program`, but mimics similar functionality,
 * including React component rendering. Params are required as they're passed to the run method,
 * while options are optional and assume class properties have been defined. Also, the `exit`
 * and `log` methods have been mocked with Vitest spies so that they can be asserted.
 *
 * ```ts
 * import { runCommand } from '@boost/cli/test';
 * import TestCommand from '../src/commands/TestCommand';
 *
 * it('runs a command', async () => {
 * 	const command = new TestCommand();
 *
 * 	expect(await runCommand(command, ['foo', 'bar', 'baz'])).toMatchSnapshot();
 * 	expect(command.log).toHaveBeenCalled();
 * });
 * ```
 *
 * > Since there is no `Program` context, any functionality that requires a program will fail. If
 * > needed, use `runProgram()` instead.
 */
export async function runCommand<O extends GlobalOptions, P extends PrimitiveType[]>(
	command: Command<O, P>,
	params: P,
	options?: Partial<O>,
): Promise<string> {
	const { vi } = await import('vitest');

	if (options) {
		Object.assign(command, options);

		// @ts-expect-error Allow overwrite
		command[INTERNAL_OPTIONS] = {
			help: false,
			locale: 'en',
			version: false,
			...options,
		};
	}

	command.exit = vi.fn();
	// eslint-disable-next-line require-atomic-updates
	command.log = await mockLogger();
	// eslint-disable-next-line require-atomic-updates
	command[INTERNAL_PARAMS] = params;

	const result = await command.run(...params);

	if (!result || typeof result === 'string') {
		return result ?? '';
	}

	return renderComponent(result);
}

/**
 * Runs a task function outside the context of a `Command`, in complete isolation.
 * A mock command context is provided with standard defaults, and can be customized
 * through the 3rd argument.
 *
 * ```ts
 * import { runTask } from '@boost/cli/test';
 * import testTask from '../src/tasks/testTask';
 *
 * it('runs a task', async () => {
 * 	const context = {
 * 		log: vi.fn(),
 * 	};
 *
 * 	expect(await runTask(testTask, ['foo', 'bar', 'baz'], context)).toMatchSnapshot();
 * 	expect(context.log).toHaveBeenCalled();
 * });
 * ```
 */
export async function runTask<A extends unknown[], R, T extends TaskContext>(
	task: (this: T, ...argz: A) => R,
	args: A,
	context?: Partial<T>,
): Promise<R> {
	const { vi } = await import('vitest');
	const notTestable = (name: string) => () => {
		throw new Error(
			`\`${name}\` is not testable using the \`runTask\` utility. Test using a full program.`,
		);
	};

	const baseContext: TaskContext = {
		exit: vi.fn(),
		help: false,
		locale: 'en',
		log: await mockLogger(),
		rest: [] as string[],
		runProgram: notTestable('runProgram'),
		runTask: notTestable('runTask'),
		unknown: {},
		version: false,
	};

	return task.apply(
		{
			...baseContext,
			...context,
		} as T,
		args,
	);
}

/**
 * Runs a `Program` as if it were ran on the command line, including middleware,
 * commands, success and failure states, and more. Utilizes mocked streams to capture
 * and return standard output and ANSI stripped output. Failed runs will not throw
 * and instead will render a failure output.
 *
 * ```ts
 * import { runProgram } from '@boost/cli/test';
 * import Program from '../src/program';
 *
 * it('runs a program', async () => {
 * 	const program = new Program();
 *
 * 	const { code, output } = await runProgram(program, ['cmd', '--foo', '123', 'bar']);
 *
 * 	expect(output).toMatchSnapshot();
 * 	expect(code).toBe(0);
 * });
 * ```
 */
export async function runProgram(
	program: Program,
	argv: string[],
	options: { append?: boolean } = {},
): Promise<{ code: ExitCode; output: string; outputStripped: string }> {
	if (!(program.streams.stderr instanceof MockWriteStream)) {
		program.streams.stderr = new MockWriteStream(options.append) as unknown as NodeJS.WriteStream;
	}

	if (!(program.streams.stdout instanceof MockWriteStream)) {
		program.streams.stdout = new MockWriteStream(options.append) as unknown as NodeJS.WriteStream;
	}

	if (!(program.streams.stdin instanceof MockReadStream)) {
		program.streams.stdin = new MockReadStream() as unknown as NodeJS.ReadStream;
	}

	// Ink async rendering never resolves while testing,
	// as it relies on system signals to "exit".
	// So we set this to flag the renderer to avoid awaiting.
	env('CLI_TEST_ONLY', 'true');

	const code = await program.run(argv);

	env('CLI_TEST_ONLY', null);

	const output =
		(program.streams.stdout as unknown as MockWriteStream).get() +
		(program.streams.stderr as unknown as MockWriteStream).get();

	return {
		code,
		output,
		outputStripped: stripAnsi(output),
	};
}
