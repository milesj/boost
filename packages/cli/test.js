const { Command } = require('./lib');

const c = new Command();
