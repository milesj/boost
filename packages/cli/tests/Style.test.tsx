import React from 'react';
import { loadTheme } from '../src/helpers/loadTheme';
import { Style } from '../src/react';
import { renderComponent } from '../src/test';
import { vi } from 'vitest';
import { describe, beforeEach, it, expect } from 'vitest';

vi.mock('term-size');
vi.mock('../src/helpers/loadTheme');

describe('<Style />', () => {
	beforeEach(() => {
		(loadTheme as jest.Mock).mockImplementation(() => ({
			default: 'white',
			failure: 'red',
			info: 'cyan',
			inverted: 'black',
			muted: 'gray',
			notice: 'magenta',
			success: 'green',
			warning: 'yellow',
		}));
	});

	it('renders `none`', async () => {
		await expect(renderComponent(<Style>Test</Style>)).resolves.toMatchSnapshot();
	});

	it('renders `none` inverted', async () => {
		await expect(renderComponent(<Style inverted>Test</Style>)).resolves.toMatchSnapshot();
	});

	it('renders `default`', async () => {
		await expect(renderComponent(<Style type="default">Test</Style>)).resolves.toMatchSnapshot();
	});

	it('renders `default` inverted', async () => {
		await expect(
			renderComponent(
				<Style inverted type="default">
					Test
				</Style>,
			),
		).resolves.toMatchSnapshot();
	});

	it('renders `info`', async () => {
		await expect(renderComponent(<Style type="info">Test</Style>)).resolves.toMatchSnapshot();
	});

	it('renders `info` inverted', async () => {
		await expect(
			renderComponent(
				<Style inverted type="info">
					Test
				</Style>,
			),
		).resolves.toMatchSnapshot();
	});

	it('renders `inverted`', async () => {
		await expect(renderComponent(<Style type="inverted">Test</Style>)).resolves.toMatchSnapshot();
	});

	it('renders `inverted` inverted', async () => {
		await expect(renderComponent(<Style type="inverted">Test</Style>)).resolves.toMatchSnapshot();
	});

	it('renders `failure`', async () => {
		await expect(renderComponent(<Style type="failure">Test</Style>)).resolves.toMatchSnapshot();
	});

	it('renders `failure` inverted', async () => {
		await expect(
			renderComponent(
				<Style inverted type="failure">
					Test
				</Style>,
			),
		).resolves.toMatchSnapshot();
	});

	it('renders `muted`', async () => {
		await expect(renderComponent(<Style type="muted">Test</Style>)).resolves.toMatchSnapshot();
	});

	it('renders `muted` inverted', async () => {
		await expect(
			renderComponent(
				<Style inverted type="muted">
					Test
				</Style>,
			),
		).resolves.toMatchSnapshot();
	});

	it('renders `notice`', async () => {
		await expect(renderComponent(<Style type="notice">Test</Style>)).resolves.toMatchSnapshot();
	});

	it('renders `notice` inverted', async () => {
		await expect(
			renderComponent(
				<Style inverted type="notice">
					Test
				</Style>,
			),
		).resolves.toMatchSnapshot();
	});

	it('renders `success`', async () => {
		await expect(renderComponent(<Style type="success">Test</Style>)).resolves.toMatchSnapshot();
	});

	it('renders `success` inverted', async () => {
		await expect(
			renderComponent(
				<Style inverted type="success">
					Test
				</Style>,
			),
		).resolves.toMatchSnapshot();
	});

	it('renders `warning`', async () => {
		await expect(renderComponent(<Style type="warning">Test</Style>)).resolves.toMatchSnapshot();
	});

	it('renders `warning` inverted', async () => {
		await expect(
			renderComponent(
				<Style inverted type="warning">
					Test
				</Style>,
			),
		).resolves.toMatchSnapshot();
	});

	it('supports bold, italics, etc', async () => {
		await expect(
			renderComponent(
				<Style bold italic type="default">
					Test
				</Style>,
			),
		).resolves.toMatchSnapshot();
	});

	// TODO: GH CI terminal does not support hexcodes
	// it('supports hexcodes', async () => {
	//   (loadTheme as jest.Mock).mockImplementation(() => ({
	//     default: '#fff',
	//     failure: 'red',
	//     inverted: '#000',
	//     muted: '#eee',
	//     success: 'green',
	//     warning: 'yellow',
	//   }));

	//   expect(await renderToString(<Style type="default">Test</Style>)).toMatchSnapshot();
	// });
});
