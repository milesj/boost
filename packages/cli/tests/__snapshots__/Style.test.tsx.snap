// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Style /> renders \`default\` 1`] = `"[37mTest[39m"`;

exports[`<Style /> renders \`default\` inverted 1`] = `"[47m[30mTest[39m[49m"`;

exports[`<Style /> renders \`failure\` 1`] = `"[31mTest[39m"`;

exports[`<Style /> renders \`failure\` inverted 1`] = `"[41m[30mTest[39m[49m"`;

exports[`<Style /> renders \`info\` 1`] = `"[36mTest[39m"`;

exports[`<Style /> renders \`info\` inverted 1`] = `"[46m[30mTest[39m[49m"`;

exports[`<Style /> renders \`inverted\` 1`] = `"[30mTest[39m"`;

exports[`<Style /> renders \`inverted\` inverted 1`] = `"[30mTest[39m"`;

exports[`<Style /> renders \`muted\` 1`] = `"[90mTest[39m"`;

exports[`<Style /> renders \`muted\` inverted 1`] = `"[100m[30mTest[39m[49m"`;

exports[`<Style /> renders \`none\` 1`] = `"Test"`;

exports[`<Style /> renders \`none\` inverted 1`] = `"Test"`;

exports[`<Style /> renders \`notice\` 1`] = `"[35mTest[39m"`;

exports[`<Style /> renders \`notice\` inverted 1`] = `"[45m[30mTest[39m[49m"`;

exports[`<Style /> renders \`success\` 1`] = `"[32mTest[39m"`;

exports[`<Style /> renders \`success\` inverted 1`] = `"[42m[30mTest[39m[49m"`;

exports[`<Style /> renders \`warning\` 1`] = `"[33mTest[39m"`;

exports[`<Style /> renders \`warning\` inverted 1`] = `"[43m[30mTest[39m[49m"`;

exports[`<Style /> supports bold, italics, etc 1`] = `"[3m[1m[37mTest[39m[22m[23m"`;
